### PR TITLE
OKTA-382172: Add IDX self hosted sample

### PR DESCRIPTION
**Pull request by arvindkrishnakumar-okta**

### CODE REVIEW

File: self-hosted/pom.xml [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>jar</packaging>
+
+    <name>Okta :: Samples :: Spring Boot :: Code Flow + Self Hosted Login Page</name>
+    <description>Okta OAuth 2.0 w/ Spring Boot Example + Self Hosted Login Page</description>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+        <com.okta.idx.sdk.version>0.1.0-beta.6-SNAPSHOT</com.okta.idx.sdk.version>
```
Comment (arvindkrishnakumar-okta) [inline self-hosted/pom.xml:37]: TODO: change snapshot to a release version.

File: self-hosted/src/main/resources/templates/login.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+</head>
+<body class="login">
+  <div id="sign-in-widget"></div>
+
+  <script th:inline="javascript">
+      /*<![CDATA[*/
+
+      var config = {};
+
+      config.baseUrl = /*[[${oktaBaseUrl}]]*/ 'https://{yourOktaDomain}';
```
Comment (JayNewstrom) [inline self-hosted/src/main/resources/templates/login.html:34]: Can we fulfill these from the environment variables we have from the backend side?

**Review by JayNewstrom — APPROVED** [review]

File: self-hosted/README.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Okta Spring Security & Self Hosted Login Page Example
+
+This example shows you how to use the [Okta Spring Boot Library][] to login a user.  The login is achieved through the Interaction Code flow using the [Okta Sign In Widget][], which gives you more control to customize the login experience within your app.  After the user authenticates, they are redirected back to the application and a local cookie session is created.
```
[the remark this answers is not in the document]
Comment (arvindkrishnakumar-okta) [reply self-hosted/README.md:3]: Yeah, the links are picked up from the bottom of the page (view it in raw mode).

File: self-hosted/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+## Prerequisites
+
+Before running this sample, you will need the following:
+
+* An Okta Developer Account, you can sign up for one at https://developer.okta.com/signup/.
+* An Okta Application, configured for Web mode. This is done from the Okta Developer Console and you can find instructions [here][OIDC Web Application Setup Instructions].  When following the wizard, use the default properties.  They are designed to work with our sample applications.
+* Your Okta Application entry needs a login redirect URI. Go to "Login redirect URIs" under "General Settings" for your application, click "Edit" and add http://localhost:8080/authorization-code/callback.
```
[the remark this answers is not in the document]
Comment (arvindkrishnakumar-okta) [reply self-hosted/README.md:12]: The link renders correctly in md format with the way it is presently. 

image

File: self-hosted/README.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+* Your Okta Application entry needs the logout callback. "Logout redirect URIs" under "General" for the application should list http://localhost:8080. If it is not present, click "Edit" and add it.
```
[the remark this answers is not in the document]
Comment (arvindkrishnakumar-okta) [reply self-hosted/README.md:13]: The link renders correctly in md format with the way it is presently.

image

File: self-hosted/src/main/java/com/okta/spring/example/CustomAuthenticationProcessingFilter.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        map.add("client_id", clientId);
+        if (Strings.hasText(clientSecret)) {
+            map.add("client_secret", clientSecret);
+        }
+        map.add("interaction_code", interactionCode);
+        map.add("code_verifier", codeVerifier);
+
+        HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(map, headers);
+        ResponseEntity<JsonNode> responseEntity =
+                restTemplate.postForEntity(issuer + "/v1/token", requestEntity, JsonNode.class);
```
[the remark this answers is not in the document]
Comment (arvindkrishnakumar-okta) [reply self-hosted/src/main/java/com/okta/spring/example/CustomAuthenticationProcessingFilter.java:175]: good catch, will update.

File: self-hosted/src/main/resources/templates/login.html (hunk 2 above)
Comment (arvindkrishnakumar-okta) [reply self-hosted/src/main/resources/templates/login.html:34]: The controller populates these values and sends it to the view. the controller uses the env variables/properties to originally fill in these values before passing it to the view.

**Review by JayNewstrom — APPROVED** [review]

File: pom.xml [hunk 7]
```diff
@@ -28,6 +28,7 @@
     <modules>
         <module>okta-hosted-login</module>
         <module>custom-login</module>
+        <module>self-hosted</module>
```
Comment (bdemers) [inline pom.xml:31]: What is the difference between "self-hosted" and "custom-login"

Comment (arvindkrishnakumar-okta) [reply pom.xml:31]: Great question! They are one and the same technically. There is some confusion on naming the new samples that would use the Interaction Code grant (OIE) and I'm not sure if I need to rename `custom-login` to `self-hosted` or host the new samples in a separate folder named `self-hosted` login. I'll clarify that with PM.

Comment (bdemers) [reply pom.xml:31]: +100 for renaming, both use the widget in a self-hosted way.  Does the original example work if you tweak the sign-in widget config?

**Comment by arvindkrishnakumar-okta** [comment]
> Can you just replace the old example with the new one?

Yes, that would be the ideal case. Awaiting product decision on this. cc @robertdamphousse-okta

**Pull request closed without merging**